### PR TITLE
Go simple and straight forward JUnit and Selenium for IT

### PR DESCRIPTION
**Pull request by GedMarc**
Using JUnit 5 and Selenium, allows any browser and any platform hosting through embedded and/or MicroService items.

Enhances the build in the following ways
- Removes test logic from the build path and allows developers to specify on which platforms and browsers the test will be executed
- Fixed the startup of tomcat and removed the maven plugin, let users and IDE's run test cases and execute the cases
- Removes any need for profiles from producing test cases
- Utilizes base JUnit 5 enhancements to manage and maintain the test cases
- Name with *IT for failsafe to pick them up, name with *Test for surefire to pick them up. No need to specify filters, rather name the test classes correctly
- Uses pure JavaScript/JQuery to confirm and perform basic test cases. WebElements and drivers still available, but not preferred.
- Adds implementation classes to provide test functionality

- Removed the primefaces selenium base artifacts and they aren't needed, but you can always put them back if you want

Includes space for adding a 3rd base functionality as well

### CODE REVIEW

**Comment by GedMarc** [comment]
2 Hour rough draft but you get the gist :)

**Comment by GedMarc** [comment]
that came across a bit pompus - just another way of doing it

**Comment by melloware** [comment]
@tandraschko @phillipross Please review and discuss.  I have no strong feelings either way but we should pick a direction.

File: src/test/java/org/primefaces/extensions/integrationtests/interactions/IDefaultTestCycle.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
+package org.primefaces.extensions.integrationtests.interactions;
+
```
Comment (melloware) [inline src/test/java/org/primefaces/extensions/integrationtests/interactions/IDefaultTestCycle.java:2]: You deleted the InputText test above but did not replace it?

**Comment by melloware** [comment]
So far I like what I see here and I like the ability to run 1 test in the IDE at a time and be able to specify which browser etc.

**Comment by tandraschko** [comment]
Didnt checked it in detail but what i dont like:
- you dont use primefaces-selenium classes (also webelements) - this is a essential for me to also test both the testclasses and PF itself; testcases are unmaintainable with just selectors and without component wrappers. therefore we introduced primefaces-selenium.
- I* prefix of interface classes
- guice support

File: src/test/java/org/primefaces/extensions/integrationtests/datepicker/DatePicker001IT.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+@Tag("DatePicker Tests")
+public class DatePicker001IT extends BaseIT {
+    private static final String datePickerInputID = "$('[id$=datepicker1_input]')";
+
+    @DisplayName("Test Input")
+    public void performTest() {
+        focus(datePickerInputID);
+        setValue(datePickerInputID, "02/19/1978");
+        blur(datePickerInputID);
+
```
Comment (melloware) [inline src/test/java/org/primefaces/extensions/integrationtests/datepicker/DatePicker001IT.java:18]: So if I am seeing this right you aren't using PrimeFaces Selenium at all you are just using Jquery?  Not sure about that we should be wrapping the Selenium complexity of a component with a wrapper like PF Selenium to hide knowing about hidden inputs etc?

Comment (phillipross) [reply src/test/java/org/primefaces/extensions/integrationtests/datepicker/DatePicker001IT.java:18]: I agree with @melloware here.  I'd like a way to wrap that complexity and do it in a reusable way.  There's a lot of tests to be written/ported and the selenium complexity needs to be manageable and maintainable.

**Comment by GedMarc** [comment]
ouch harsh 🗡️ :) 
Yeah there were no interfaces to look at in the repo's as a reference ;) Wanted to highlight a clear distinction between WebElement and IWebElement
I explained my reasoning behind not using WebElement and going JQuery selectors, as well as a full example on gitter :)

The guice thing is my throw in, was highlighting that there isn't a limit to how expandable on either runtime or browser or device and that you can make your integration tests simple and effective no matter how deep or complex they become :)

Again, this is just another way of doing it :) It really all falls down to the ecosystem that you know - its not possible for me to know those intricate details. All code is opinionated :)

File: src/test/java/org/primefaces/extensions/integrationtests/datepicker/DatePicker001IT.java (hunk 2 above)
Comment (GedMarc) [reply src/test/java/org/primefaces/extensions/integrationtests/datepicker/DatePicker001IT.java:18]: Yeah :) I find it still the cleanest way, You test visibility, text content, structure, and layout in a single call, also I kind of reckon you're testing html more than the java really in integration tests, sometimes i suppose.

A nice thing that this also does though, is allows you to return data objects from the page, as Expression<>, so you can return ```JSON.parse(PF['remoteVar'].data.cfg);``` (Which I do frequently) and de-serialize for validations

So I go JQuery 3 with CSS3 Selectors - https://www.w3schools.com/cssref/css_selectors.asp - very powerful.

File: src/test/java/org/primefaces/extensions/integrationtests/interactions/IWebActionable.java [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+		String output = (String) ((JavascriptExecutor) getDriver()).executeScript(runnable);
+		return output;
+	}
+
+	default boolean doesExist(String jquerySelector)
+	{
+		//Wait for jQuery to load
+		JavascriptExecutor jsExec = (JavascriptExecutor) getDriver();
+		WebDriverWait jsWait = new WebDriverWait(getDriver(), 10);
+		ExpectedCondition<Boolean> jQueryLoad = driver -> ((Long) ((JavascriptExecutor) getDriver())
```
Comment (GedMarc) [inline src/test/java/org/primefaces/extensions/integrationtests/interactions/IWebActionable.java:109]: Returning type safe arguments from the javascript engine in the browser -> PhantomJS compatible :) @melloware

File: src/test/java/org/primefaces/extensions/integrationtests/interactions/IWebActionable.java [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+	 * Waits 10 seconds for an element to be available, or throws a failed exception
+	 *
+	 * @param cssSearchSelector
+	 */
+	default void waitForElement(String cssSearchSelector) {
+		focusOnWindow();
+		WebDriverWait wait = new WebDriverWait(getDriver(), 10);
+		try {
+			System.out.println("Waiting for  : " + cssSearchSelector + "  : To be visible");
+			WebElement element = wait.until(ExpectedConditions.visibilityOfElementLocated(By.cssSelector(cssSearchSelector)));
```
Comment (GedMarc) [inline src/test/java/org/primefaces/extensions/integrationtests/interactions/IWebActionable.java:134]: @tandraschko WebElement wait ()

File: src/test/java/org/primefaces/extensions/integrationtests/interactions/IWebActionable.java [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+			wait.withMessage("ready").until(a -> false);
+		}catch (Throwable T)
+		{
+			//intentional
+		}
+	}
+
+	default void focusOnWindow()
+	{
+		((JavascriptExecutor)getDriver()).executeScript("window.focus(); intermediate = function () {window.requestAnimationFrame(fn)}\n"
```
Comment (GedMarc) [inline src/test/java/org/primefaces/extensions/integrationtests/interactions/IWebActionable.java:166]: nb for firefox testing.
If you start an integration test with firefox and you click away, all tests will fail

**Pull request closed without merging**